ext/opcache/jit/zend_jit_arm64.dasc

```diff
@@ -0,0 +1,6050 @@
+|.define TMP3w,   w10
+|.define TMP4,    x11
+|.define TMP4w,   w11
+|.define FPTMP1,  v16
+|.define FPTMP2,  v17
+
+// Temporary register index in _zend_reg
+|.define ZREG_TMP1,   ZREG_X8
+|.define ZREG_TMP2,   ZREG_X9
+|.define ZREG_TMP3,   ZREG_X10
+|.define ZREG_TMP4,   ZREG_X11
+|.define ZREG_FPTMP1, ZREG_V16
+|.define ZREG_FPTMP2, ZREG_V17
+
+#define ZREG_TMP1     ZREG_X8
+#define ZREG_TMP2     ZREG_X9
+#define ZREG_TMP3     ZREG_X10
+#define ZREG_TMP4     ZREG_X11
+#define ZREG_FPTMP1   ZREG_V16
+#define ZREG_FPTMP2   ZREG_V17
+
+|.define HYBRID_SPAD, #16     // padding for stack alignment
+
+#define TMP_ZVAL_OFFSET 0
+#define DASM_ALIGNMENT  16
+#define MAX_IMM12 0xfff       // maximum value for imm12
+
+#include "Zend/zend_cpuinfo.h"
+
+#ifdef HAVE_VALGRIND
+# include <valgrind/valgrind.h>
+#endif
+
+/* The generated code may contain tautological comparisons, ignore them. */
+#if defined(__clang__)
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wtautological-compare"
+# pragma clang diagnostic ignored "-Wstring-compare"
+#endif
+
+const char* zend_reg_name[] = {
+	"x0", "x1", "x2", "x3", "x4", "x5", "x6", "x7",
+	"x8", "x9", "x10", "x11", "x12", "x13", "x14", "x15",
+	"x16", "x17", "x18", "x19", "x20", "x21", "x22", "x23",
+	"x24", "x25", "x26", "x27", "x28", "x29", "x30", "sp",
+	"v0", "v1", "v2", "v3", "v4", "v5", "v6", "v7",
+	"v8", "v9", "v10", "v11", "v12", "v13", "v14", "v15",
+	"v16", "v17", "v18", "v19", "v20", "v21", "v22", "v23",
+	"v24", "v25", "v26", "v27", "v28", "v29", "v30", "v31"
+};
+
+#ifdef HAVE_GCC_GLOBAL_REGS
+# define GCC_GLOBAL_REGS 1
+#else
+# define GCC_GLOBAL_REGS 0
+#endif
+
+# define ZREG_FCARG1x ZREG_X0
+# define ZREG_FCARG2x ZREG_X1
+
+#if ZTS
+static size_t tsrm_ls_cache_tcb_offset = 0;
+static size_t tsrm_tls_index;
+static size_t tsrm_tls_offset;
+#endif
+
+/* By default avoid JITing inline handlers if it does not seem profitable due to lack of
+ * type information. Disabling this option allows testing some JIT handlers in the
+ * presence of try/catch blocks, which prevent SSA construction. */
+#ifndef PROFITABILITY_CHECKS
+# define PROFITABILITY_CHECKS 1
+#endif
+
+|.type EX, zend_execute_data, FP
+|.type OP, zend_op
+|.type ZVAL, zval
+
+|.actionlist dasm_actions
+
+|.globals zend_lb
+static void* dasm_labels[zend_lb_MAX];
+
+|.section code, cold_code, jmp_table
+
+#define IS_32BIT(addr) (((uintptr_t)(addr)) <= 0x7fffffff)
+
+#define IS_SIGNED_32BIT(val) ((((intptr_t)(val)) <= 0x7fffffff) && (((intptr_t)(val)) >= (-2147483647 - 1)))
+
+#define BP_JIT_IS 6
+
+/* In x86/64, HYBRID_SPAD bytes are reserved on the stack only if flag ZEND_VM_HYBRID_JIT_RED_ZONE_SIZE
+ * is not defined, because the 16-byte redzone, allocated on the stack when the flag is defined, can be
+ * reused. In AArch64, it's safe that these bytes are always reserved because the stack layout might
+ * change along software evolution, making the redzone not reusable any longer. */
+|.macro ADD_HYBRID_SPAD
```

Comment:
Then you don't need redsone. It was introduced to save this add/sub in each trace.

Reply:
@dstogov Thanks for your review. Yes. We didn't use redzone here.
As I see it, the primary goal that redzone was introduced is to prevent 'vm_stack_data' from being optimized out by compilers. I got this from https://github.com/php/php-src/pull/6398. And the secondary goal of redzone I think is to be reused here, i.e. saving the add/sub.
Hence, I suppose redzone should always be defined for AArch64 (See updates in file zend_vm_opcdoes.h), and we may need decide whether we use it for saving the add/sub for ADD_HYBRID_SPAD. I think option 1) we don't reuse it. As I mentioned in the comments, it would be simple and safe, and option 2) we reuse it as the x86 implementation does. Do you think which on is better? Thanks.

Reply:
I use redzone to optimize out add/sub at each JIT code entry/exit. Better to keep the code similar to x86 and support both possibilities, dependent on \#define.

Reply:
Thanks for your explanation. Will update.